iter/iter_test.go

```diff
@@ -166,6 +168,312 @@ func TestForEach(t *testing.T) {
+		t.Parallel()
+
+		wantConcurrency := 2 * defaultMaxGoroutines()
+
+		maxConcurrencyHit := make(chan struct{})
+
+		tasks := make([]int, wantConcurrency)
+		iterator := Iterator[int]{MaxGoroutines: wantConcurrency}
+
+		var concurrentTasks atomic.Int64
+		_ = iterator.ForEachErr(tasks, func(t *int) error {
+			n := concurrentTasks.Add(1)
+			defer concurrentTasks.Add(-1)
+
+			if int(n) == wantConcurrency {
+				// All our tasks are running concurrently.
+				// Signal to the rest of the tasks to stop.
+				close(maxConcurrencyHit)
+			} else {
+				// Wait until we hit max concurrency before exiting.
+				// This ensures that all tasks have been started
+				// in parallel, despite being a larger input set than
+				// defaultMaxGoroutines().
+				<-maxConcurrencyHit
+			}
+
+			return nil
+		})
+	})
+}
+
+func TestForEachIdxErr(t *testing.T) {
+	t.Parallel()
+
+	t.Run("standart", func(t *testing.T) {
+		forEach := noIndex(ForEachIdxErr[int])
+		testForEachErr(t, false, forEach)
+	})
+
+	t.Run("unique indexes", func(t *testing.T) {
+		ints := []int{1, 2, 3, 4, 5}
+		got := []int{}
+		gotMu := sync.Mutex{}
+
+		err := ForEachIdxErr(ints, func(i int, _ *int) error {
+			gotMu.Lock()
+			defer gotMu.Unlock()
+			got = append(got, i)
+			return nil
+		})
+
+		require.NoError(t, err)
+		require.ElementsMatch(t, got, []int{0, 1, 2, 3, 4})
+	})
+}
+
+func TestForEachErr(t *testing.T) {
+	t.Parallel()
+
+	testForEachErr(t, false, ForEachErr[int])
+}
+
+// noIndex converts a ForEachIdxErr function (or method) into a ForEachErr function (or method).
+func noIndex(forEach func([]int, func(int, *int) error) error) func([]int, func(*int) error) error {
+	return func(ints []int, fn func(*int) error) error {
+		return forEach(ints, func(_ int, val *int) error {
+			return fn(val)
+		})
+	}
+}
+
+// testForEachErr runs a set of tests against a ForEachErr function (or method).
+func testForEachErr(t *testing.T, failFast bool, forEach func([]int, func(*int) error) error) {
+	t.Run("empty", func(t *testing.T) {
+		t.Parallel()
+		f := func() {
+			ints := []int{}
+			_ = forEach(ints, func(val *int) error {
+				panic("this should never be called")
+			})
+		}
+		require.NotPanics(t, f)
+	})
+
+	t.Run("panic is propagated", func(t *testing.T) {
+		t.Parallel()
+		f := func() {
+			ints := []int{1}
+			_ = forEach(ints, func(val *int) error {
+				panic("super bad thing happened")
+			})
+		}
+		require.Panics(t, f)
+	})
+
+	t.Run("mutating inputs is fine", func(t *testing.T) {
+		t.Parallel()
+		ints := []int{1, 2, 3, 4, 5}
+		_ = forEach(ints, func(val *int) error {
+			*val += 1
+			return nil
+		})
+		require.Equal(t, []int{2, 3, 4, 5, 6}, ints)
+	})
+
+	t.Run("mutating inputs is fine", func(t *testing.T) {
+		t.Parallel()
+		ints := []int{1, 2, 3, 4, 5}
+		_ = forEach(ints, func(val *int) error {
+			*val += 1
+			return nil
+		})
+		require.Equal(t, []int{2, 3, 4, 5, 6}, ints)
+	})
```

Comment:
Looks like a duplicate
```suggestion
	t.Run("mutating inputs is fine", func(t *testing.T) {
		t.Parallel()
		ints := []int{1, 2, 3, 4, 5}
		_ = forEach(ints, func(val *int) error {
			*val += 1
			return nil
		})
		require.Equal(t, []int{2, 3, 4, 5, 6}, ints)
	})
```

Reply:
fixed